src/main/java/tech/stackable/hadoop/StackableTopologyProvider.java

```diff
@@ -175,144 +164,193 @@ private String getLabel(
-    LOG.debug("Resolved nodelabels for: " + nodeLabels.keySet().toString());
-    Map<String, Map<String, String>> podLabels = getPodLabels(pods);
-    LOG.debug("Resolved podlabels for " + podLabels.keySet().toString());
+
+    Map<String, Map<String, String>> nodeLabels = getNodeLabels(datanodes);
+    LOG.debug("Resolved node labels map [{}]/[{}]", nodeLabels.keySet(), nodeLabels.values());
+
+    Map<String, Map<String, String>> podLabels = getPodLabels(datanodes);
+    LOG.debug("Resolved pod labels map [{}]/[{}]", podLabels.keySet(), podLabels.values());
+
+    names = resolveDataNodesFromCallingPods(names, podLabels, datanodes);
 
     // Iterate over all nodes to resolve and return the topology zones
-    for (String datanode : names) {
-      String builtLabel = getLabel(datanode, podLabels, nodeLabels);
+    for (String pod : names) {
+      String builtLabel = getLabel(pod, podLabels, nodeLabels);
       result.add(builtLabel);
 
       // Cache the value for potential use in a later request
-      this.topologyKeyCache.put(datanode, builtLabel);
+      this.topologyKeyCache.put(pod, builtLabel);
     }
+    LOG.info("Returning resolved labels [{}]", result);
     return result;
   }
 
   /**
-   * Given a list of Pods, return a HashMap that maps pod ips onto Pod labels. The returned Map may
-   * contain more entries than the list that is given to this function, as an entry will be
+   * The list of names may be datanodes (as is the case when the topology is initialised) or
+   * non-datanodes, when the data is being written by a client tool, spark executors etc. In this
+   * case we want to identify the datanodes that are running on the same node as this client. The
+   * names may also be pod IPs or pod names.
+   *
+   * @param names list of client pods to resolve to datanodes
+   * @param podLabels map of podIPs and labels
+   * @param dns list of datanode nameswhich will be used to match nodenames
+   * @return
+   */
+  private List<String> resolveDataNodesFromCallingPods(
+      List<String> names, Map<String, Map<String, String>> podLabels, List<Pod> dns) {
+    List<String> dataNodes = new LinkedList<>();
+    for (String name : names) {
+      // if we don't find a dataNode running on the same node as a non-dataNode pod, then
+      // we'll keep the original name to allow it to be resolved to NotFound in the calling routine.
+      String replacementDataNodeIp = name;
+      InetAddress address;
+      try {
+        // make sure we are looking up using the IP address
+        address = InetAddress.getByName(name);
+        String podIp = address.getHostAddress();
+        if (podLabels.containsKey(podIp)) {
+          replacementDataNodeIp = podIp;
+          LOG.info("added as found in the datanode map [{}]", podIp);
+        } else {
+          // we've received a call from a non-datanode pod
+          List<Pod> pods = client.pods().list().getItems();
```

Comment:
I think this talks to the k8s api and potentially fetches quite a bit of data, does this have to be inside of the `for (String name : names) {` loop?
Or maybe initialize pods with "null" outside of the loop and only retrieve pods if its "null" here, so we can avoid the call alltogether if we can answer everything from our cached values.

Just to double check, any work we do here will be written to the cache and be cached for a while, correct?

Reply:
Done [3025fcc](https://github.com/stackabletech/hdfs-topology-provider/pull/5/commits/3025fcc1298c84c20a830d8fb1bb48e8bf00e564). 

> Just to double check, any work we do here will be written to the cache and be cached for a while, correct?

No not yet, but I would rather do this in a separate issue when we re-work the cache to have a keyset of all pod IPs rather than just the DataNodes.

Reply:
I've created an issue here: https://github.com/stackabletech/hdfs-topology-provider/issues/6

Reply:
Could we maybe do a cheap'n cheerful caching along the general lines of the idea I have written down in this commit?

https://github.com/stackabletech/hdfs-topology-provider/commit/d1231358ffad2e82bda1d1eb656f4825d2b985ad

Reply:
Good idea! Done: b412818

Reply:
And also: a9e1dbd